From 4838eeb486df7bce075da97ea1233935b6688a2e Mon Sep 17 00:00:00 2001
From: Jan Silhan <jsilhan@redhat.com>
Date: Mon, 13 Jun 2016 23:05:44 +0200
Subject: [PATCH] subject: prefer obsoletes (RhBug:1096506)(RhBug:1332830)

DNF makes better selection of packages considering obsoletes from
'pkg_spec' input when `multilib_policy` is set to `best`
(by Subject.set_best_selector()). Moreover the selection of packages
is evaluated during resolution so when the obsoleting package has
broken dependencies the next best choose is picked (RhBug:1096506)

Subject.get_best_selector searches by provides first (RhBug:1332830)
while prioritizing obsoletes and then packages with the same name as a
provide.

Closes: #609
Approved by: j-mracek
---
 dnf/subject.py | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/dnf/subject.py b/dnf/subject.py
index ea7a0e57..a1b3fbf4 100644
--- a/dnf/subject.py
+++ b/dnf/subject.py
@@ -123,9 +123,14 @@ class Subject(object):
         nevra = first(self.subj.nevra_possibilities_real(sack, **kwargs))
         if nevra:
             sltr = dnf.selector.Selector(sack)
-            s = self._nevra_to_selector(sltr, nevra)
-            if len(s.matches()) > 0:
-                return s
+            if nevra._has_just_name():
+                s = sltr.set(provides=nevra.name)
+                if len(s.matches()) > 0:
+                    return s
+            else:
+                s = self._nevra_to_selector(sltr, nevra)
+                if len(s.matches()) > 0:
+                    return s
 
         reldep = first(self.subj.reldep_possibilities_real(sack))
         if reldep:
-- 
2.11.0

